ngraph/core/include/ngraph/runtime/reference/interpolate.hpp

```diff
@@ -0,0 +1,633 @@
+                                    std::min(coord, static_cast<int64_t>(length) - 1));
+                }
+
+                float triangle_coeff(float dz) { return std::max(0.0f, 1.0f - std::fabs(dz)); }
+                std::array<float, 4> get_cubic_coeff(float s, float a)
+                {
+                    std::array<float, 4> coeff;
+                    float abs_s = std::fabs(s);
+                    coeff[0] = static_cast<float>(
+                        ((a * (abs_s + 1) - 5 * a) * (abs_s + 1) + 8 * a) * (abs_s + 1) - 4 * a);
+                    coeff[1] = static_cast<float>(((a + 2) * abs_s - (a + 3)) * abs_s * abs_s + 1);
+                    coeff[2] = static_cast<float>(
+                        ((a + 2) * (1 - abs_s) - (a + 3)) * (1 - abs_s) * (1 - abs_s) + 1);
+                    coeff[3] = static_cast<float>(
+                        ((a * (2 - abs_s) - 5 * a) * (2 - abs_s) + 8 * a) * (2 - abs_s) - 4 * a);
+                    return coeff;
+                }
+
+                float get_in_coord(float coord, int64_t axis_idx)
+                {
+                    float scale = m_scales[axis_idx];
+                    int64_t axis = m_axes[axis_idx];
+                    float length_resized = static_cast<float>(m_out_shape[axis]);
+                    float length_original = static_cast<float>(m_input_data_shape[axis]);
+                    return m_get_original_coord(coord, scale, length_resized, length_original);
+                }
+            };
+
+            template <typename T>
+            void InterpolateEval<T>::linear_func(const T* input_data, T* out)
+            {
+                std::size_t input_rank = m_input_data_shape.size();
+                std::size_t num_of_axes = m_axes.size();
+                bool is_downsample = false;
+                for (std::size_t scale : m_scales)
+                {
+                    is_downsample = is_downsample || (scale < 1.0);
+                }
+
+                bool antialias = is_downsample && m_antialias;
+
+                std::vector<float> a(num_of_axes);
+                std::vector<int64_t> r(num_of_axes);
+
+                CoordinateTransform output_transform(m_out_shape);
+                CoordinateTransform input_transform(m_input_data_shape);
+
+                std::vector<int64_t> low_limits_vector(num_of_axes);
+                std::vector<int64_t> high_limits_vector(num_of_axes);
+                float prod_a = 1;
+                for (std::size_t i = 0; i < num_of_axes; ++i)
+                {
+                    a[i] = antialias ? m_scales[i] : 1.0;
+                    prod_a *= a[i];
+                    r[i] = (m_scales[i] > 1.0) ? static_cast<int64_t>(2)
+                                               : static_cast<int64_t>(std::ceil(2.0f / a[i]));
+                    low_limits_vector[i] = -r[i];
+                    high_limits_vector[i] = r[i];
+                }
+
+                runtime::NDimIndex minimal_indices{
```

Comment:
Why do you need `NDimIndex` here?

I am sure that all classes `NDim*` are redundant and you can use existed nGraph classes for same things.

Reply:
Now I am rewriting this code in `CoordinateTransform`.